source/common/http/http1/codec_impl.cc

```diff
@@ -411,8 +411,9 @@ int ServerConnectionImpl::onHeadersComplete(HeaderMapImplPtr&& headers) {
         http_parser_pause(&parser_, 1);
       }
 
-    } else if (parser_.method == http_method::HTTP_POST ||
-               parser_.method == http_method::HTTP_PUT) {
+    } else if ((parser_.method == http_method::HTTP_POST ||
+                parser_.method == http_method::HTTP_PUT) &&
+               headers->ContentLength() == nullptr) {
```

Comment:
Yeah was going to link that. Also https://tools.ietf.org/html/rfc7230#section-3.3.3: 

"6.  If this is a request message and none of the above are true, then
       the message body length is zero (no message body is present)."

IMO as long as http-parser correctly infers content-length: 0 if none is set for request, we should probably not do this check. We should possible insert content-length: 0 for higher layers in this case?